Fix reference to a released stack variable in modem_context_sprint_ip…

ssharks opened the pull request:
Suggested fix for issue #38172. How can I test it?

Fixes #38172.

ssharks commented:
Fix the build errors

stephanosio commented:
Please review the [Contribution Guidelines](https://docs.zephyrproject.org/latest/contribute/index.html#commit-guidelines) and re-format the commit messages.

ssharks commented:
I hope this a correct comment message

rerickson1 reviewed (changes requested):
1. `drivers: modem:` repeated twice in you commit message.
2. Need a description body in your commit message
3. Need Signed-off-by in your commit body.

ssharks committed:
drivers: modem: Fix reference to stack in modem_context_sprint_ip_addr

Changes the modem_context_sprint_ip addr to write into a provided
buffer. This approach fixes a referencing to a non existing stack
variable, without using a lot of stack space.
Tested using the by building all used modems and tested on HW using
IPv4.

Fixes: #38459

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

cfriedt commented:
@ssharks - just some whitespace changes in your commit message for CI to pass. Please modify with `commit --amend` and do a force push

stephanosio commented on drivers/modem/modem_context.c:
```diff
@@ -20,23 +20,37 @@ LOG_MODULE_REGISTER(modem_context, CONFIG_MODEM_LOG_LEVEL);
 	LOG_ERR("Unknown IP address family:%d", addr->sa_family);
-	strcpy(buf, "unk");
-	return buf;
+	char unknown_str[] = "unk";
```
Consider moving variable declarations to the top of the scope (function).

stephanosio commented on drivers/modem/quectel-bg9x.c:
```diff
@@ -735,9 +735,20 @@ static int offload_connect(void *obj, const struct sockaddr *addr,
 
 	k_sem_reset(&mdata.sem_sock_conn);
 
+	char ip_str[NET_IPV6_ADDR_LEN];
```
Consider moving variable declarations to the top of the scope (function).

stephanosio commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -1574,8 +1582,17 @@ static int offload_connect(void *obj, const struct sockaddr *addr,
 		return 0;
 	}
 
+	char ip_str[NET_IPV6_ADDR_LEN];
```
Consider moving variable declarations to the top of the scope (function).

stephanosio commented on drivers/modem/modem_context.c:
```diff
@@ -20,23 +20,38 @@ LOG_MODULE_REGISTER(modem_context, CONFIG_MODEM_LOG_LEVEL);
+			return -ENOMEM;
+		}
+
+		net_addr_ntop(AF_INET6, &net_sin6(addr)->sin6_addr,
+				     buf, buf_size);
+		buf[buf_size - 1] = '\0';
```
This should not be necessary; instead, it should check the return value of `net_addr_ntop` and return an error code upon failure in which case the content of `buf` should not be used.

ssharks replied:
Seems reasonable I just overlooked this review comment, processed it

stephanosio commented on drivers/modem/modem_context.c:
```diff
@@ -20,23 +20,38 @@ LOG_MODULE_REGISTER(modem_context, CONFIG_MODEM_LOG_LEVEL);
+int modem_context_sprint_ip_addr(const struct sockaddr *addr, char *buf, int buf_size)
 {
-	static char buf[NET_IPV6_ADDR_LEN];
+	char unknown_str[] = "unk";
```
```suggestion
	const char unknown_str[] = "unk";
```

ssharks replied:
Done

stephanosio commented:
> Does anyone has an idea why the maintainers check failed this time?

@ssharks Try rebasing onto the latest master. A few entries in the `MAINTAINERS.yml` were fixed recently.

stephanosio commented on drivers/modem/modem_context.h:
```diff
@@ -91,10 +91,12 @@ struct modem_context {
+ * @retval 0 if ok, < 0 if error.
  */
-char *modem_context_sprint_ip_addr(const struct sockaddr *addr);
+int modem_context_sprint_ip_addr(const struct sockaddr *addr, char *buf, int buf_size);
```
```suggestion
int modem_context_sprint_ip_addr(const struct sockaddr *addr, char *buf, size_t buf_size);
```

ssharks replied:
Good suggestions, I implemented them

ssharks commented on drivers/modem/modem_context.c:
```diff
@@ -20,23 +20,38 @@ LOG_MODULE_REGISTER(modem_context, CONFIG_MODEM_LOG_LEVEL);
+		}
+		net_addr_ntop(AF_INET, &net_sin(addr)->sin_addr,
+				     buf, buf_size);
+		buf[buf_size - 1] = '\0';
```
Same

ssharks commented:
@
 
Collaborator
cfriedt ### Thanks for reminding, just overlooked it

stephanosio commented on drivers/modem/quectel-bg9x.c:
```diff
@@ -735,9 +736,18 @@ static int offload_connect(void *obj, const struct sockaddr *addr,
 	/* Formulate the complete string. */
 	snprintk(buf, sizeof(buf), "AT+QIOPEN=%d,%d,\"%s\",\"%s\",%d,0,0", 1, sock->sock_fd, protocol,
-		 modem_context_sprint_ip_addr(addr), dst_port);
+			ip_str, dst_port);
```
Nit (styling issue): the beginning of this line is no longer aligned with the `(` from the previous line.
```suggestion
		 ip_str, dst_port);
```

ssharks replied:
Fixed

stephanosio commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -1574,8 +1583,15 @@ static int offload_connect(void *obj, const struct sockaddr *addr,
 		return 0;
 	}
 
+	ret = modem_context_sprint_ip_addr(addr, ip_str, sizeof(ip_str));
+	if (ret != 0) {
+		errno = -ret;
+		LOG_ERR("Error formatting IP string %d", ret);
+		return -1;
+	}
+
 	snprintk(buf, sizeof(buf), "AT+USOCO=%d,\"%s\",%d", sock->id,
-		 modem_context_sprint_ip_addr(addr), dst_port);
+			ip_str, dst_port);
```
same here

ssharks replied:
Fixed, I also remove the spurious space

stephanosio replied:
Thanks. The space is not spurious though -- without the space, it is no longer properly aligned:
```
	snprintk(buf, sizeof(buf), "AT+USOCO=%d,\"%s\",%d", sock->id,
		 modem_context_sprint_ip_addr(addr), dst_port);
		ip_str, dst_port);
```

cfriedt replied:
Looks to be resolved, based on the Coding Guidelines checks.

stephanosio reviewed (commented):
Looks ok overall.
